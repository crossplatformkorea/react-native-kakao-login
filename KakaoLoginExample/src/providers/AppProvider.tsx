import React, {useReducer} from 'react';

import {User} from '../types';
import createCtx from '../utils/createCtx';

interface Context {
  state: State;
  setUser: (user: User) => void;
  resetUser: () => void;
  callDefault: () => void;
}

const [useCtx, Provider] = createCtx<Context>();

export enum ActionType {
  ResetUser = 'reset-user',
  SetUser = 'set-user',
  CallDefault = 'call-default',
}

export interface State {
  user: User | null;
}

const initialState: State = {
  user: null,
};

interface SetUserAction {
  type: ActionType.SetUser;
  payload: User;
}

interface ResetUserAction {
  type: ActionType.ResetUser;
}

interface GetStateAction {
  type: ActionType.CallDefault;
}

type Action = SetUserAction | ResetUserAction | GetStateAction;

interface Props {
  children?: React.ReactElement;
}

type Reducer = (state: State, action: Action) => State;

const callDefault = (dispatch: React.Dispatch<GetStateAction>) => (): void => {
  dispatch({
    type: ActionType.CallDefault,
  });
};

const setUser = (dispatch: React.Dispatch<SetUserAction>) => (
  user: User,
): void => {
  dispatch({
    type: ActionType.SetUser,
    payload: user,
  });
};

const resetUser = (dispatch: React.Dispatch<ResetUserAction>) => (): void => {
  dispatch({
    type: ActionType.ResetUser,
  });
};

// eslint-disable-next-line default-param-last
const reducer: Reducer = (state = initialState, action) => {
  switch (action.type) {
    case 'reset-user':
      return initialState;
    case 'set-user':
      return {...state, user: action.payload};
    default:
      return state;
  }
};

function AppProvider(props: Props): React.ReactElement {
  const [state, dispatch] = useReducer<Reducer>(reducer, initialState);

  const actions = {
    setUser: setUser(dispatch),
    resetUser: resetUser(dispatch),
    callDefault: callDefault(dispatch),
  };

  return <Provider value={{state, ...actions}}>{props.children}</Provider>;
}

export {useCtx as useAppContext, AppProvider};
