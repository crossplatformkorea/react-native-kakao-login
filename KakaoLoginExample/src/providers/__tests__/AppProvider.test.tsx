import * as React from 'react';

import {AppProvider, useAppContext} from '../AppProvider';
import {Button, Text, View} from 'react-native';
import {
  RenderAPI,
  act,
  cleanup,
  fireEvent,
  render,
} from '@testing-library/react-native';

let testingLib: RenderAPI;

const FakeChild = (): React.ReactElement => {
  const {state, resetUser, callDefault} = useAppContext();

  return (
    <View>
      <Text testID="TEXT">{JSON.stringify(state, null, 2)}</Text>
      <Button
        testID="BUTTON"
        onPress={(): void => {
          resetUser();
        }}
        title="Button"
      />
      <Button
        testID="BUTTON_NOT_VALID"
        onPress={(): void => {
          callDefault();
        }}
        title="Button"
      />
    </View>
  );
};

describe('[AppProvider] rendering test', () => {
  const component = (
    <AppProvider>
      <FakeChild />
    </AppProvider>
  );

  afterEach(cleanup);

  it('should match component and snapshot', () => {
    testingLib = render(component);

    const baseElement = testingLib.toJSON();

    expect(baseElement).toMatchSnapshot();
    expect(baseElement).toBeTruthy();
  });

  it('should call [resetUser] when button pressed', () => {
    testingLib = render(component);

    const btn = testingLib.queryByTestId('BUTTON');

    act(() => {
      fireEvent.press(btn);
    });
  });

  it('should call [default] when button pressed', () => {
    testingLib = render(component);

    const btn = testingLib.queryByTestId('BUTTON_NOT_VALID');

    act(() => {
      fireEvent.press(btn);
    });
  });
});

describe('[AppProvider] error rendering test', () => {
  let error: Error;
  const component = <FakeChild />;

  afterEach(cleanup);

  it('should throw error when [AppProvider] is not wrapped', () => {
    try {
      render(component);
    } catch (e) {
      error = e;
    }

    expect(error).toBeInstanceOf(Error);
  });
});
