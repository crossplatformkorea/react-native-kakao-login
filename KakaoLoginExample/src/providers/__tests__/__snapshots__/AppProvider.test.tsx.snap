// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[AppProvider] rendering test should match component and snapshot 1`] = `
<View>
  <Text
    testID="TEXT"
  >
    {
  "user": null
}
  </Text>
  <View
    accessibilityRole="button"
    accessibilityState={Object {}}
    onPress={[Function]}
    testID="BUTTON"
  >
    <View
      style={
        Array [
          Object {},
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#007AFF",
              "fontSize": 18,
              "margin": 8,
              "textAlign": "center",
            },
          ]
        }
      >
        Button
      </Text>
    </View>
  </View>
  <View
    accessibilityRole="button"
    accessibilityState={Object {}}
    onPress={[Function]}
    testID="BUTTON_NOT_VALID"
  >
    <View
      style={
        Array [
          Object {},
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#007AFF",
              "fontSize": 18,
              "margin": 8,
              "textAlign": "center",
            },
          ]
        }
      >
        Button
      </Text>
    </View>
  </View>
</View>
`;
